resolves ScottishCovidResponse/SCRCIssueTracking#296

anthonyohare opened the pull request:
Added docstrings for the functions in network_of_populations.py in reStructuredText format.

bobturneruk commented:
Thank you so much @anthonyohare! I hope @aflag or @magicicada will be able to have a look at it as they can much more quickly verify the docstrings than I can.

bobturneruk commented:
This seems very picky as I write it, and **I should've put it in the issue**, but ideally this would pass a [`pydocstyle`](https://pypi.org/project/pydocstyle/) linting check. Thanks again, though. This definitely should not stop a merge! This looks like great progress to me. More of a note to myself.

bobturneruk commented:
@anthonyohare - my attention has been drawn to some conflicts with the current `master` branch. I think the best way to deal with these would be for you to run the following:
```
git fetch
git checkout issue-296
git rebase origin/master
git push --force
```
This will apply your changes as if on top of the current latest `master` (in which there are a slightly different set of functions in `simple_network_sim/network_of_populations.py`). This may (but hopefully won't) reveal some new undocumented functions.

If you'd prefer me to do this, please let me know.

anthonyohare commented:
Thanks @bobturneruk, pydocstyle doesn't recognise the sphinx format unfortunately (it only recognises google, numpy and pep257). I've rebased with origin/master and added new comments for new parameters.

bobturneruk commented:
Thanks for that @anthonyohare! My thoughts with `pydocstyle` were mostly just that it helps identify which bits have a docstring and helps a bit with consistency. If @aflag or @magicicada could jump in and have a look at this, that would be great as they are more familiar with this code than me (having written it!). If not I'll have a look as soon as I can. Thanks again for doing this.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -14,6 +14,17 @@
 
 # CurrentlyInUse
 def countInfectiousAgeStructured(dictOfStates, time):
+    """Count the number of infectious individuals in all states at some time.
```
I think instead of *all states* it should be *all nodes and across all age groups* or something like that

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -14,6 +14,17 @@
+    :type dictOfStates: A dictionary with time as keys and whose values are another dictionary with
+    the region as a key and the disease state as values. The states are a dictionary with a tuple
+    of (age, state) as keys and the number of individuals in that state as values.
+    :param time: The time (day) within the simulation.
```
I'd leave (day) out. The data doesn't need to be based on days (though it currently is). As long as they all agree, the model wouldn't care.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -83,10 +100,26 @@ def basicSimulationInternalAgeStructure(network, timeHorizon):
 
 # CurrentlyInUse
 def totalIndividuals(nodeState):
+    """Calculate the total population size over all ages and regions.
```
This function takes a node (region) and counts individuals from every age and state.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -83,10 +100,26 @@ def basicSimulationInternalAgeStructure(network, timeHorizon):
 def totalIndividuals(nodeState):
+    """Calculate the total population size over all ages and regions.
+
+    :param nodeState: The disease status of the population stratified by age.
```
I think you should at least mention that this is the data from a node (or region).

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -83,10 +100,26 @@ def basicSimulationInternalAgeStructure(network, timeHorizon):
 
 
 def getAges(node):
+    """Get the unique set of ages from the node.
```
A set is already unique

anthonyohare replied:
This was sloppy of me; I meant "unique collection of" but set is a better word.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -83,10 +100,26 @@ def basicSimulationInternalAgeStructure(network, timeHorizon):
+    :type node: A dictionary with a tuple of (age, state) as keys and the number of individuals
+    in that state as values.
+    :return: The unique collection of ages.
+    :rtype: Set of str
```
Maybe `Set[str]`?

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -95,6 +128,16 @@ def getAges(node):
 
 # CurrentlyInUse
 def getTotalInAge(nodeState, ageTest):
+    """Get the size of the population within an age range.
```
I know the class is called "age range" but I think saying "age group" in the documentation would be clearer.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -104,6 +147,14 @@ def getTotalInAge(nodeState, ageTest):
 
 # CurrentlyInUse
 def getTotalInfectious(nodeState):
+    """Get the total number of infected individuals regardless of age.
```
It should be "infectious" and not "infected". There are people who are infected, but not infectious (eg. exposed and hospitalized).

Also, it's worth mentioning this function does that for a specific node.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -112,16 +163,34 @@ def getTotalInfectious(nodeState):
 
 # CurrentlyInUse
 def getTotalSuscept(nodeState):
+    """Get the total number of susceptible individuals regardless of age.
```
I'd mention "in the node" here

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -139,6 +208,19 @@ def distributeInfections(nodeState, newInfections):
 
 
 def doIncomingInfectionsByNode(graph, currentState, movementMultiplier):
+    """Determine the number of new infections at each node of a graph.
```
I'd add something like *based on incoming people from neighboring nodes* just to make it extra clear that this function deals with between nodes infections.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -160,14 +242,23 @@ def doIncomingInfectionsByNode(graph, currentState, movementMultiplier):
 def getWeight(graph, orig, dest, multiplier):
+    """Get the weight of the edge from orig to dest in the graph. If the edge doesn't have a
+    weight, 1.0 is assumed and the returned weight is adjusted by the multiplier and any
+    delta_adjustment on the edge.
```
I think this function name is a bit too generic, but maybe worth mentioning in the docs that this weight is expected to be proportional to the movement between nodes.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -139,6 +208,19 @@ def distributeInfections(nodeState, newInfections):
 def doIncomingInfectionsByNode(graph, currentState, movementMultiplier):
+    """Determine the number of new infections at each node of a graph.
+
+    :param graph: A graph with each region as a node and the weights corresponding to the commutes
```
I'd replace the word commute with movement or something else more generic. I think we are only using commute data because that's what we have, but ideally we'd use all the movements among regions.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -194,6 +285,24 @@ def getWeight(graph, orig, dest, multiplier):
 #  We may need to multiply movement numbers by a probability of infection to achieve this.   
 def getExternalInfections(graph, dictOfStates, currentTime, movementMultiplier):
+    """Calculate the number of new infections in each region, distributed across each age group
+    relative to the size [of the number of susceptibles] of each age group.
```
I think there's two much information in the same phrase. Perhaps: *Calculate the number of new infections in each region. The infections are distributed proportionally to the number of susceptibles in the destination node and infected in the origin node. The infections are distributed to each age group according to the number of suscepitible people in them.*

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -213,6 +322,20 @@ def getExternalInfections(graph, dictOfStates, currentTime, movementMultiplier):
 #  and the probability of each of these being infectious is 0.25, then I would expect the matrix going into this
 # function as  ageMixingInfectionMatrix to have 0.3 in the entry [age1][age2]
 def doInternalInfectionProcess(currentInternalStateDict, ageMixingInfectionMatrix, contactsMultiplier):
+    """Calculate the new infections due to mixing with the region and stratify them by age.
```
Did you mean *within*?

anthonyohare replied:
yes, thanks.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -251,11 +393,21 @@ def getInternalInfection(dictOfStates, ageMixingInfectionMatrix, time, contactsM
-#  So we will need to do some accounting
 def internalStateDiseaseUpdate(currentInternalStateDict, diseaseProgressionProbs):
+    """Update the status of exposed individuals, moving them into the next disease state with a
+    probability defined in the given progression matrix.
```
I think it's worth mentioning that this function returns a dictionary with new state, rather than updating things in-place.

aflag commented on simple_network_sim/network_of_populations.py:
```diff
@@ -489,23 +709,46 @@ def createNextStep(progression, exposed, currState):
 
 
 def getSusceptibles(age, currentInternalStateDict):
+    """Calculate the total number of individuals in a susceptible state in an age range.
```
Maybe worth mentioning that the second parameter is a region

bobturneruk commented:
Thanks for the review here @aflag! And thanks again for the hard work on this @anthonyohare! 

I think once this is accurate in that it reflects what the code does, it should go in. If time is limited, improvements to wording could happen later as a separate issue.

anthonyohare commented:
Thanks for the comprehensive review @aflag. I'll do these now.
